primitives/manta/src/assets.rs

```diff
@@ -306,210 +374,267 @@ impl FungibleLedgerError {
         }
     }
 
+    /// Non-native assets will use the `can_increase_total_supply` flag, while native assets will
+    /// not.
     #[inline]
     fn can_deposit(
-        asset_id: AssetId,
+        asset_id: Self::AssetId,
         account: &C::AccountId,
-        amount: Balance,
+        amount: Self::Balance,
         can_increase_total_supply: bool,
-    ) -> Result<(), FungibleLedgerError> {
-        Self::ensure_valid(asset_id)?;
+    ) -> Result<Self::AssetId, FungibleLedgerError<Self::AssetId, Self::Balance>> {
+        let asset_id = Self::ensure_valid(asset_id)?;
         FungibleLedgerError::from_deposit(if asset_id == A::NativeAssetId::get() {
-            <Native as FungibleInspect<C::AccountId>>::can_deposit(account, amount, false)
+            Native::can_deposit(account, amount, false)
         } else {
-            <NonNative as FungiblesInspect<C::AccountId>>::can_deposit(
-                asset_id,
-                account,
-                amount,
-                can_increase_total_supply,
-            )
+            NonNative::can_deposit(asset_id.clone(), account, amount, can_increase_total_supply)
         })
+        .map(|_| asset_id)
     }
 
+    /// Will mint and increase the total supply of non-native assets.
     #[inline]
-    fn can_reduce_by_amount(
-        asset_id: AssetId,
+    fn deposit_minting(
+        asset_id: Self::AssetId,
         account: &C::AccountId,
-        amount: Balance,
-        existence_requirement: ExistenceRequirement,
-    ) -> Result<(), FungibleLedgerError> {
-        Self::ensure_valid(asset_id)?;
-
-        let keep_alive = match existence_requirement {
-            ExistenceRequirement::KeepAlive => true,
-            ExistenceRequirement::AllowDeath => false,
-        };
-        let reducible_amount = if asset_id == A::NativeAssetId::get() {
-            <Native as FungibleInspect<C::AccountId>>::reducible_balance(account, keep_alive)
+        amount: Self::Balance,
+    ) -> Result<(), FungibleLedgerError<Self::AssetId, Self::Balance>> {
+        let asset_id = Self::ensure_valid(asset_id)?;
+        if asset_id == A::NativeAssetId::get() {
+            Native::deposit_creating(account, amount);
         } else {
-            <NonNative as FungiblesInspect<C::AccountId>>::reducible_balance(
-                asset_id, account, keep_alive,
-            )
-        };
-
-        if reducible_amount >= amount {
-            return Ok(());
+            NonNative::mint_into(asset_id, account, amount)
+                .map_err(FungibleLedgerError::InvalidMint)?;
         }
-        Err(FungibleLedgerError::CannotWithdrawMoreThan(
-            reducible_amount,
-        ))
+        Ok(())
     }
 
     #[inline]
-    fn deposit_can_mint(
-        asset_id: AssetId,
-        beneficiary: &C::AccountId,
-        amount: Balance,
-    ) -> Result<(), FungibleLedgerError> {
-        Self::ensure_valid(asset_id)?;
+    fn try_deposit_minting(
```

Comment:
technically, both `deposit` and `try_deposit` are fallible and return result, the `try` naming is misleading

Reply:
or `deposit_minting_with_check`

Reply:
Personally I prefer the explicit naming like `deposit_minting_with_check` so I'll change it and revert if @bhgomes feels strongly about the try naming